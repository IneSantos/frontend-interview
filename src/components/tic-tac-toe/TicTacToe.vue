<template>
  <div class="tictactoe-board">
    <div class="grid">
      <div v-for="(row, row_index) in 3" :key="row">
        <div v-for="(column, column_index) in 3" :key="column">
          <cell :value="board[row_index][column_index]"  @click="playMove(row_index,column_index)"/>
        </div>
      </div>
    </div>

    <div class="score">
      <Player name="Player1" :score="player1"/>

      <Player name="Player2" :score="player2"/>
    </div>
  </div>
</template>

<script>
import '@/styles/tic-tac-toe/tictactoe.scss';
import Player from './Player';

export default {
  name: 'TicTacToe',
  components: {
    Player
  },
  data() { 
    return {
      board:  [
        ['', '', ''],
        ['', '', ''],
        ['', '', '']
      ],
      player: "x",
      player1: 0,
      player2: 0,
      finished: false,
      stalemate: false
    }
  },
  methods: {
      playMove(x, y) {
        if (this.finished || this.board[x][y] !== '') {
          // Invalid move.
          return;
        }

        this.board[x][y] = this.player;
        this.$forceUpdate();

        this.finished = this.checkWinner();
        this.stalemate = this.checkStalemate();
        this.updateScore();
        this.nextPlayer();
        
      },
      nextPlayer: function() {
        this.player = this.player === "x" ? "o" : "x";
      },  
      checkStalemate: function() {
        return (
          !this.finished &&
          this.checkValuesPresent(this.board[0]) &&
          this.checkValuesPresent(this.board[1]) &&
          this.checkValuesPresent(this.board[2])
        );
      },
      checkWinner: function() {
        return (
          this.checkValues(this.board[0]) ||
          this.checkValues(this.board[1]) ||
          this.checkValues(this.board[2]) ||
          this.checkValues([this.board[0][0], this.board[1][0], this.board[2][0]]) ||
          this.checkValues([this.board[0][1], this.board[1][1], this.board[2][1]]) ||
          this.checkValues([this.board[0][2], this.board[1][2], this.board[2][2]]) ||
          this.checkValues([this.board[0][0], this.board[1][1], this.board[2][2]]) ||
          this.checkValues([this.board[0][2], this.board[1][1], this.board[2][0]])
        );
      },
      checkValues: function(values) {
        return this.checkValuesPresent(values) && this.checkValuesMatch(values);
      },
      checkValuesPresent: function(values) {
        return values[0] !== "" && values[1] !== "" && values[2] !== "";
      },
      checkValuesMatch: function(values) {
        return values[0] === values[1] && values[1] === values[2];
      },
      updateScore: function() {
        if(this.finished) return;
        this.player === "x" ? ++this.player1 : ++this.player2;
      }

  }

}
</script>