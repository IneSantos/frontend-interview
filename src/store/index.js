export const store = {
    state: {
      hasgameStarted: false,
      numOfgames: 0,
      scorePlayer1: 0,
      scorePlayer2: 0,
      winsPlayer1: 0,
      winsPlayer2: 0,
      lossPlayer1: 0,
      lossPlayer2:0,
      gameHistory: ['', '', '' , '', '']
    },
    mutations: {
      SET_GAME_STARTED(state, hasgameStarted) {
        state.hasgameStarted = hasgameStarted
      },
      INC_NUM_GAMES(state) {
        state.numOfgames++
      },
      SET_SCORE_PLAYER_1(state, scorePlayer1) {
        state.scorePlayer1 = scorePlayer1
      },
      SET_SCORE_PLAYER_2(state, scorePlayer2) {
        state.scorePlayer2 = scorePlayer2
      },
      SET_WINS_PLAYER_1(state, winsPlayer1) {
        state.winsPlayer1 = winsPlayer1
      },
      SET_WINS_PLAYER_2(state, winsPlayer2) {
        state.winsPlayer2 = winsPlayer2
      },
      SET_LOSS_PLAYER_1(state, lossPlayer1) {
        state.lossPlayer1 = lossPlayer1
      },
      SET_LOSS_PLAYER_2(state, lossPlayer2) {
        state.lossPlayer2 = lossPlayer2
      },
      SET_GAME_HISTORY(state, player) {
        state.push(player);
      },
      RESET_GAME(state) {
        state.hasgameStarted = false;
      }
    },
    actions: {
      setGameStart(context, hasgameStarted){
        context.commit('SET_GAME_STARTED', hasgameStarted);
      },
      resetBoard(context){
        context.commit('RESET_BOARD')
      },
      incNumOfGames(context){
        context.commit('INC_NUM_GAMES');
      },
      setScorePlayer1(context, scorePlayer1){
        context.commit('SET_SCORE_PLAYER_1', scorePlayer1);
      },
      setScorePlayer2(context, scorePlayer2){
        context.commit('SET_SCORE_PLAYER_2', scorePlayer2);
      },
      setWinsPlayer1(context, winsPlayer1){
        context.commit('SET_WINS_PLAYER_1', winsPlayer1);
      },
      setWinsPlayer2(context, winsPlayer2){
        context.commit('SET_WINS_PLAYER_2', winsPlayer2);
      },
      setLossPlayer1(context, lossPlayer1) {
        context.commit('SET_LOSS_PLAYER_1', lossPlayer1);
      },
      setLossPlayer2(context, lossPlayer2) {
        context.commit('SET_LOSS_PLAYER_2', lossPlayer2);
      },
      setGameHistory(context, player){
        context.commit('SET_GAME_HISTORY', player);
      }
    },
    getters: {
      getHasGameStarted(state) {
        return state.hasgameStarted;
      },
      getNumOfGames(state) {
        return state.numOfgames;
      },
      getScorePlayer1(state) {
        return state.scorePlayer1;
      },
      getScorePlayer2(state) {
        return state.scorePlayer2;
      },
      getWinsPlayer1(state) {
        return state.winsPlayer1;
      },
      getWinsPlayer2(state) {
        return state.winsPlayer2;
      },
      getLossPlayer1(state) {
        return state.lossPlayer1;
      },
      getLossPlayer2(state) {
        return state.lossPlayer2;
      },
      getGameHistory(state) {
        return state.gameHistory;
      },
    }

}

