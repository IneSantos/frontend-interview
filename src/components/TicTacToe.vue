
<template>
  <section class="games">
    <h1>Tic tac toe games</h1>
    <p>Welcome to the best game in the world.</p>
    <GamesToggle @toggle="toggleGames" ref="gameToggle"/>
    <Board v-if="showTicTacToe"/>
    <Board4inRow v-else/>
  </section>
</template>


<script>
import '@/styles/tic-tac-toe.scss';
import Board4inRow from './4-in-a-row/Board4inRow'
import Board from './tic-tac-toe/Board';
import GamesToggle from './GamesToggle';

export default {
  name: 'TicTacToe',
  components: {
    Board,
    Board4inRow,
    GamesToggle
  }, 
  data() {
    return {
      showTicTacToe: true,
    };
  },
  methods: {
      toggleGames(showTicTacToe) {
          this.showTicTacToe = showTicTacToe;
      },
  }
}
</script>