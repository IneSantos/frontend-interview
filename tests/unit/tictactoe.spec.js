import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import TicTacToe from '@/components/TicTacToe.vue'
import { storeMocked } from '../__mocks__/store.js'

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store(storeMocked);

describe('TicTacToe.vue', () => {
  it('renders default', () => {
    const wrapper = shallowMount(TicTacToe);
    expect(wrapper).toMatchSnapshot();
  })

  it('Should should call toggle when toggle button is clicked', async () => {
    const mockMethod = jest.spyOn(TicTacToe.methods, 'toggleGames');
    const wrapper = mount( TicTacToe, { store, localVue } );

    await wrapper.findComponent({ ref: "gameToggle" }).vm.$emit('toggle');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should should change option shown when toggle button is clicked', async () => {
    const wrapper = mount( TicTacToe, { store, localVue } );

    expect(wrapper.vm.$data.showTicTacToe).toBeTruthy();
    await wrapper.findComponent({ ref: "gameToggle" }).vm.$emit('toggle');
    expect(wrapper.vm.$data.showTicTacToe).toBeFalsy();
  })

})
