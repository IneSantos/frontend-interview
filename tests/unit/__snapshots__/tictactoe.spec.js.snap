// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTacToe.vue renders default 1`] = `
<section class="games">
  <h1>Tic tac toe games</h1>
  <p>Welcome to the best game in the world.</p>
  <gamestoggle-stub></gamestoggle-stub>
  <board-stub></board-stub>
</section>
`;
